File: reference.vb.net/Db4objects.Db4odoc/CommitCallbacks/PushedUpdatesExample.vb
' Copyright (C) 2007 db4objects Inc. http://www.db4o.com
Imports System
Imports System.IO
Imports System.Threading
Imports System.Collections
Imports Db4objects.Db4o
Imports Db4objects.Db4o.Events
Imports Db4objects.Db4o.Ext
Imports Db4objects.Db4o.Foundation

Namespace Db4objects.Db4odoc.CommitCallbacks

    Class PushedUpdatesExample
        Private Const Db4oFileName As String = "reference.db4o"
        Private Const Port As Integer = 4440
        Private Const User As String = "db4o"
        Private Const Password As String = "db4o"
        Private Shared clientListeners As Hashtable = New Hashtable

        Public Shared Sub Main(ByVal args As String())
            Call (New PushedUpdatesExample).Run()
        End Sub
        ' end Main

        Public Sub Run()
            File.Delete(Db4oFileName)
            Dim server As IObjectServer = Db4oFactory.OpenServer(Db4oFileName, Port)
            Try
                server.GrantAccess(User, Password)
                Dim client1 As IObjectContainer = OpenClient
                Dim client2 As IObjectContainer = OpenClient
                If Not (client1 Is Nothing) AndAlso Not (client2 Is Nothing) Then
                    Try
                        ' wait for the operations to finish
                        WaitForCompletion()

                        'save pilot with client1
                        Dim client1Car As Car = New Car("Ferrari", 2006, New Pilot("Schumacher"))
                        client1.Set(client1Car)
                        client1.Commit()
                        WaitForCompletion()

                        ' retrieve the same pilot with client2
                        Dim client2Car As Car = CType(client2.Query(GetType(Car)).Next, Car)
                        System.Console.WriteLine(client2Car)

                        ' modify the pilot with pilot1
                        client1Car.Model = 2007
                        client1Car.Pilot = New Pilot("Hakkinnen")
                        client1.Set(client1Car)
                        client1.Commit()
                        WaitForCompletion()

                        ' client2Car has been automatically updated in
                        ' the committed event handler because of the
                        ' modification and the commit by client1
                        System.Console.WriteLine(client2Car)
                        WaitForCompletion()
                    Catch ex As Exception
                        System.Console.WriteLine(ex.ToString)
                    Finally
                        CloseClient(client1)
                        CloseClient(client2)
                    End Try
                End If
            Catch ex As Exception
                System.Console.WriteLine(ex.ToString)
            Finally
                server.Close()
            End Try
        End Sub
        ' end Run

        Private Sub CloseClient(ByVal client As IObjectContainer)
            ' remove listeners before shutting down
            If Not (clientListeners(client) Is Nothing) Then
                Dim eventRegistry As IEventRegistry = EventRegistryFactory.ForObjectContainer(client)
                RemoveHandler eventRegistry.Committed, AddressOf CType(clientListeners(client), CommittedEventHandler).OnCommitted
                clientListeners.Remove(client)
            End If
            client.Close()
        End Sub
        ' end CloseClient

        Private Function OpenClient() As IObjectContainer
            Try
                Dim client As IObjectContainer = Db4oFactory.OpenClient("localhost", Port, User, Password)
                Dim committedEventHandler As CommittedEventHandler = New CommittedEventHandler(client)
                Dim eventRegistry As IEventRegistry = EventRegistryFactory.ForObjectContainer(client)
                AddHandler eventRegistry.Committed, AddressOf committedEventHandler.OnCommitted
                ' save the client-listener pair in a map, so that we can
                ' remove the listener later
                clientListeners.Add(client, committedEventHandler)
                Return client
            Catch ex As Exception
                System.Console.WriteLine(ex.ToString)
            End Try
            Return Nothing
        End Function
        ' end OpenClient

        Private Sub WaitForCompletion()
            Try
                Thread.Sleep(1000)
            Catch e As ThreadInterruptedException
                System.Console.WriteLine(e.Message)
            End Try
        End Sub
        ' end WaitForCompletion

    End Class
End Namespace

File: reference/reference.net/db4o-reference-chapters/src/Db4objects.Db4odoc/CommitCallbacks/PushedUpdatesExample.cs
using System;
using System.IO;
using System.Threading;
using System.Collections;

using Db4objects.Db4o;
using Db4objects.Db4o.Events;
using Db4objects.Db4o.Ext;
using Db4objects.Db4o.Foundation;

namespace Db4objects.Db4odoc.CommitCallbacks
{
    class PushedUpdatesExample
    {

        private const string Db4oFileName = "reference.db4o";

        private const int Port = 4440;

        private const string User = "db4o";

        private const string Password = "db4o";

        private static Hashtable clientListeners = new Hashtable();

        public static void Main(string[] args)
        {
            new PushedUpdatesExample().Run();
        }

        // end Main

        public void Run()
        {
            File.Delete(Db4oFileName);
            IObjectServer server = Db4oFactory.OpenServer(Db4oFileName, Port);
            try
            {
                server.GrantAccess(User, Password);

                IObjectContainer client1 = OpenClient();
                IObjectContainer client2 = OpenClient();

                if (client1 != null && client2 != null)
                {
                    try
                    {
                        // wait for the operations to finish
                        WaitForCompletion();

                        // save pilot with client1
                        Car client1Car = new Car("Ferrari", 2006, new Pilot("Schumacher"));
                        client1.Set(client1Car);
                        client1.Commit();

                        WaitForCompletion();

                        // retrieve the same pilot with client2
                        Car client2Car = (Car)client2.Query(typeof(Car)).Next();
                        System.Console.WriteLine(client2Car);

                        // modify the pilot with client1
                        client1Car.Model = 2007;
                        client1Car.Pilot = new Pilot("Hakkinnen");
                        client1.Set(client1Car);
                        client1.Commit();

                        WaitForCompletion();

                        // client2Car has been automatically updated in
                        // the committed event handler because of the
                        // modification and the commit by client1
                        System.Console.WriteLine(client2Car);

                        WaitForCompletion();
                    }
                    catch (Exception ex)
                    {
                        System.Console.WriteLine(ex.ToString());
                    }
                    finally
                    {
                        CloseClient(client1);
                        CloseClient(client2);
                    }
                }
            }
            catch (Exception ex)
            {
                System.Console.WriteLine(ex.ToString());
            }
            finally
            {
                server.Close();
            }
        }

        // end Run

        private void CloseClient(IObjectContainer client)
        {
            // remove listeners before shutting down
            if (clientListeners[client] != null)
            {
                IEventRegistry eventRegistry = EventRegistryFactory.ForObjectContainer(client);
                eventRegistry.Committed -= (CommitEventHandler)clientListeners[client];
                clientListeners.Remove(client);
            }
            client.Close();
        }

        // end CloseClient

        private IObjectContainer OpenClient()
        {
            try
            {
                IObjectContainer client = Db4oFactory.OpenClient("localhost", Port, User,
                        Password);
                CommitEventHandler committedEventHandler = CreateCommittedEventHandler(client);
                IEventRegistry eventRegistry = EventRegistryFactory.ForObjectContainer(client);
                eventRegistry.Committed += committedEventHandler;
                // save the client-listener pair in a map, so that we can
                // remove the listener later
                clientListeners.Add(client, committedEventHandler);
                return client;
            }
            catch (Exception ex)
            {
                System.Console.WriteLine(ex.ToString());
            }
            return null;
        }

        // end OpenClient

        private CommitEventHandler CreateCommittedEventHandler(IObjectContainer objectContainer)
        {
            return new CommitEventHandler(delegate(object sender, CommitEventArgs args)
                {
                    // get all the updated objects
                    IObjectInfoCollection updated = args.Updated;

                    foreach (IObjectInfo info in updated)
                    {
                        Object obj = info.GetObject();
                        // refresh object on the client
                        objectContainer.Ext().Refresh(obj, 2);
                    }
                });
        }

        // end CreateCommittedEventHandler

        private void WaitForCompletion()
        {
            try
            {
                Thread.Sleep(1000);
            }
            catch (ThreadInterruptedException e)
            {
                System.Console.WriteLine(e.Message);
            }
        }
        // end WaitForCompletion

    }
}
